### PR TITLE
upstream: add phantom types to load/availability

### LINKED ISSUE #5570: Stronger typing in load_balancer_impl
With the work on degraded lb (https://github.com/envoyproxy/envoy/pull/5367) the code now has a lot of different `std::vector<uint32_t>` types: degraded/healthy load, degraded/healthy availability. The lack of typing here makes working on this code somewhat painful, as passing the wrong type leads to subtle errors that take a while to debug. 

It would be nice to improve the typing here to make the compiler do the work to enforce that the right variables are used. 

One option is to use phantom types for this, eg:

```
template<class T, class... Ts>
struct Phantom : T {
	template<class... S>
	explicit Phantom(S&&... v) : T(std::forward<S>(v)...) {}
};

struct Load {};
typedef Phantom<std::vector<uint32_t>, Load> PriorityLoad;

struct Health {};
struct Degraded {};

typedef Phantom<PriorityLoad, Health> HealthPriorityLoad;
typedef Phantom<PriorityLoad, Degraded> DegradedPriorityLoad;

void priority(PriorityLoad pl) {}
void healthPriority(HealthPriorityLoad hl) {}

int main() {
	HealthPriorityLoad hpl{1, 2};
	priority(hpl); // works because HealthPriorityLoad : PriorityLoad
	healthPriority(hpl);
	PriorityLoad pl{3,4};
	priority(pl);
	healthPriority(hpl); // doesn't work since PriorityLoad isn't a HealthPriorityLoad
	return 0;
}
```
(https://ideone.com/h7E7MO)

this would allow using `HealthPriorityLoad` for more generic types that use `PriorityLoad` for reusable functions while still being able to differentiate it from `DegradedPriorityLoad` for more specific functions. 

Using the wrong type yields an error like:

```
prog.cpp:26:8: error: could not convert ‘pl’ from ‘PriorityLoad {aka Phantom<std::vector<unsigned int>, Load>}’ to ‘HealthPriorityLoad {aka Phantom<Phantom<std::vector<unsigned int>, Load>, Health>}’
  bar(pl);

``` 
using gcc.

Open for other suggestions as well, just figured I'd start out with a concrete proposal.

### PR DESCRIPTION
**Pull request by snowp**
This adds a `Phantom` template that can be used to add additional type information to
other types, which allows for stronger typing between types represented by the same
type. Applies this to `PriorityLoad` and `PriorityAvailability`, both of which are represented
by a `std::vector<uint32_t>`.

*Risk Level*: Low
*Testing*: UTs for `Phantom`, everything else is compile time
*Docs Changes*: n/a
*Release Notes*: n/a
Fixes #5570

### CODE REVIEW

**Comment by snowp** [comment]
This might be overkill, as the `Phantom` template is more generic than it has to be than if we implemented it just for `std::vector<T>`. Happy to dumb it down a bit if necessary.

This approach means that the phantom types can be used interchangeably from the type it
wraps, at the cost of being a bit trickier to initialize than an approach like https://www.fluentcpp.com/2016/12/08/strong-types-for-strong-interfaces/, but avoids having to call `get()` on the type to get ahold of the inner value.

File: source/common/common/phantom.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include <initializer_list>
+#include <type_traits>
+#include <utility>
+
+namespace Envoy {
+
+template <class I, class M> struct Phantom;
+
+// Minor optimization: keeping these as free functions avoids duplicating them for every
+// Phantom type, at the cost of making the globally accessible.
```
Comment (mattklein123) [inline source/common/common/phantom.h]: s/the/them

**Comment by mattklein123** [comment]
> This approach means that the phantom types can be used interchangeably from the type it
wraps, at the cost of being a bit trickier to initialize than an approach like https://www.fluentcpp.com/2016/12/08/strong-types-for-strong-interfaces/, but avoids having to call get() on the type to get ahold of the inner value.

FWIW, I find this type of code to be incredibly hard to read and understand and would prefer to avoid it if possible (though it is quite impressive!). Did you consider just make a struct with an embedded vector and passing the struct around where applicable? Yes this would be most verbose for using code, though you could do something like overload `operator()` to make getting the inner vector simpler? Thoughts? @envoyproxy/maintainers any opinions here?

**Comment by snowp** [comment]
Yeah, I think the complexity of the generic implementation here is probably not worth it. I think I'll move over to using a generic wrapper like you suggested (that's how the linked blog post does it), and then if I can figure out an elegant generic solution I can revisit this.

**Comment by htuch** [comment]
@mattklein123 I'll probably find myself in an unpopular minority and say I like the pseudo-functional style advocated in the linked article and use of strong/phantom types here. They are basically doing ADTs in C++. This is largely due to a bias towards type safety over all else. Readability is a wash; you get greater clarity at value creation time but there is a lot of `get()` pollution, so meh.

I don't know if it's a huge win for the specific use case being advocated here, so count my opinion as from the peanut gallery.

File: source/common/common/phantom.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <utility>
+
+namespace Envoy {
+
+// A phantom type allows wrapping a common type with addtional type information in order to allow
+// additional compile time safety when passing it around.
+template <class InnerT, class TagT> struct Phantom {
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/common/phantom.h]: No I would leave it generic. I like this.

File: source/common/common/phantom.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+namespace Envoy {
+
+// A phantom type allows wrapping a common type with additional type information in order to allow
+// additional compile time safety when passing it around.
+template <class InnerT, class TagT> struct Phantom {
+  Phantom() = default;
+  explicit Phantom(const InnerT& t) : val_(t) {}
+  explicit Phantom(InnerT&& t) : val_(std::move(t)) {}
+
+  InnerT& get() { return val_; }
```
Comment (mattklein123) [inline source/common/common/phantom.h]: As I mentioned previously, if you want it to be even more succinct, you could consider overloading `operator()` instead of `get()`. Personally, I prefer this since it's less magical, but just throwing it out there.

**Comment by snowp** [comment]
@htuch In case you missed it I previously had an approach up that used a bunch of metaprogramming to attempt to allow for a phantom type without having to call `get()` with the same type restrictions as the current approach, which I think is what Matt referred to when he talked about code being hard to read. At this point I think that was the wrong way to do it due to the implementation complexity.

I definitely think the additional type safety of the current approach (which is basically the one in the blog I linked) is worth the additional clutter of `get()`s. I have some thoughts on how to improve this, but I'd rather do that in subsequent PR, and then we can compare implementation vs implementation directly. For now I think this easy to understand approach is the way to go.

I dumped some thoughts on different phantom implementations here https://medium.com/@snowp/transparent-phantom-types-in-c-de6ac5bed1d1 in case anyone is interested in getting stuck in a template metaprogramming rabbit hole.

File: source/common/common/phantom.h (hunk 3 above)
Comment (snowp) [reply source/common/common/phantom.h]: Yeah I agree on making this less magical, `operator()` makes it look like a function object

**Comment by htuch** [comment]
@mattklein123 @snowp yeah, I was looking at the second version. So, we are all in agreement then :) The blog post makes it a bit clearer what was going on before, I also find that hard to read.
